File: apps/memory-lab-py/memory_lab_py/extraction_skill.py
from __future__ import annotations

import re
from datetime import datetime

from .utils_text import normalize_text, truncate


def _extract_from_pattern(text: str, pattern: re.Pattern[str]) -> list[str]:
    results: list[str] = []
    for match in pattern.finditer(text):
        value = normalize_text(match.group(1) if match.lastindex else "")
        if value:
            results.append(value)
    return results


def extract_project_tags(messages: list[dict], skills: dict) -> list[str]:
    tags: list[str] = []
    seen: set[str] = set()
    user_text = "\n".join([str(msg.get("content", "")) for msg in messages if str(msg.get("role", "")) == "user"])
    extraction_rules = skills["extractionRules"]
    for pattern in extraction_rules["projectPatterns"]:
        for value in _extract_from_pattern(user_text, pattern):
            cleaned = re.sub(r"[^\w.-]", "", value)
            if (
                len(cleaned) >= extraction_rules["projectTagMinLength"]
                and len(cleaned) <= extraction_rules["projectTagMaxLength"]
                and cleaned not in seen
            ):
                seen.add(cleaned)
                tags.append(cleaned)
    return tags[: extraction_rules["maxProjectTags"]]


def extract_fact_candidates(messages: list[dict], skills: dict) -> list[dict]:
    facts: dict[str, dict] = {}
    user_text = "\n".join([str(msg.get("content", "")) for msg in messages if str(msg.get("role", "")) == "user"])
    extraction_rules = skills["extractionRules"]
    for rule in extraction_rules["factRules"]:
        for value in _extract_from_pattern(user_text, rule["regex"]):
            text = truncate(value, rule["maxLength"])
            key = f"{rule['keyPrefix']}:{text.lower()}"
            facts[key] = {
                "factKey": key,
                "factValue": text,
                "confidence": float(rule["confidence"]),
            }

    for project_name in extract_project_tags(messages, skills):
        key = f"project:{project_name.lower()}"
        facts[key] = {
            "factKey": key,
            "factValue": project_name,
            "confidence": 0.78,
        }
    return list(facts.values())[: extraction_rules["maxFacts"]]


def build_session_summary(messages: list[dict], skills: dict) -> str:
    user_messages = [normalize_text(str(msg.get("content", ""))) for msg in messages if str(msg.get("role", "")) == "user"]
    assistant_messages = [
        normalize_text(str(msg.get("content", ""))) for msg in messages if str(msg.get("role", "")) == "assistant"
    ]

    user_head = user_messages[0] if user_messages else ""
    user_tail = user_messages[-1] if user_messages else ""
    assistant_tail = assistant_messages[-1] if assistant_messages else ""
    limits = skills["extractionRules"]["summaryLimits"]

    parts: list[str] = []
    if user_head:
        parts.append(f"用户提到：{truncate(user_head, limits['head'])}")
    if user_tail and user_tail != user_head:
        parts.append(f"后续重点：{truncate(user_tail, limits['tail'])}")
    if assistant_tail:
        parts.append(f"助手响应：{truncate(assistant_tail, limits['assistant'])}")
    if not parts:
        return "该窗口没有可用文本，跳过结构化摘要。"
    return "；".join(parts)


def build_situation_time_info(timestamp: str, summary: str) -> str:
    try:
        dt = datetime.fromisoformat(timestamp.replace("Z", "+00:00"))
    except Exception:
        return f"未知时间场景：{truncate(summary, 120)}"
    return f"{dt.date().isoformat()} {dt.strftime('%H:%M')} 用户正在推进：{truncate(summary, 120)}"

File: packages/memory-core/src/skills/extraction-skill.ts
import type { FactCandidate, MemoryMessage } from "../types.js";
import { normalizeText, truncate } from "../utils/text.js";

const PROJECT_PATTERNS = [
  /(?:项目|project)\s*[:：]?\s*([A-Za-z][A-Za-z0-9_-]{1,48})/gi,
  /\b([A-Z][A-Za-z0-9]+(?:[A-Z][A-Za-z0-9]+)+)\b/g,
];

const FACT_PATTERNS: Array<{ regex: RegExp; keyPrefix: string; confidence: number }> = [
  { regex: /(?:我在用|我使用|使用的是|技术栈是)\s*([A-Za-z0-9.+#_-]{2,40})/gi, keyPrefix: "tech", confidence: 0.82 },
  { regex: /(?:我正在|我在)\s*([^，。,.!?]{2,60})/gi, keyPrefix: "activity", confidence: 0.68 },
  { regex: /(?:喜欢|偏好)\s*([^，。,.!?]{2,40})/gi, keyPrefix: "preference", confidence: 0.72 },
  { regex: /(?:计划|准备)\s*([^，。,.!?]{2,40})/gi, keyPrefix: "plan", confidence: 0.65 },
];

function extractFromPattern(text: string, pattern: RegExp): string[] {
  const results: string[] = [];
  for (const match of text.matchAll(pattern)) {
    const value = normalizeText(match[1] ?? "");
    if (value) results.push(value);
  }
  return results;
}

export function extractProjectTags(messages: MemoryMessage[]): string[] {
  const tags = new Set<string>();
  const userText = messages
    .filter((msg) => msg.role === "user")
    .map((msg) => msg.content)
    .join("\n");

  for (const pattern of PROJECT_PATTERNS) {
    for (const value of extractFromPattern(userText, pattern)) {
      const cleaned = value.replace(/[^\w.-]/g, "");
      if (cleaned.length >= 2 && cleaned.length <= 50) {
        tags.add(cleaned);
      }
    }
  }
  return Array.from(tags).slice(0, 8);
}

export function extractFactCandidates(messages: MemoryMessage[]): FactCandidate[] {
  const facts = new Map<string, FactCandidate>();
  const userText = messages
    .filter((msg) => msg.role === "user")
    .map((msg) => msg.content)
    .join("\n");

  for (const { regex, keyPrefix, confidence } of FACT_PATTERNS) {
    for (const value of extractFromPattern(userText, regex)) {
      const text = truncate(value, 120);
      const key = `${keyPrefix}:${text.toLowerCase()}`;
      facts.set(key, {
        factKey: key,
        factValue: text,
        confidence,
      });
    }
  }

  for (const projectName of extractProjectTags(messages)) {
    facts.set(`project:${projectName.toLowerCase()}`, {
      factKey: `project:${projectName.toLowerCase()}`,
      factValue: projectName,
      confidence: 0.78,
    });
  }

  return Array.from(facts.values()).slice(0, 16);
}

export function buildSessionSummary(messages: MemoryMessage[]): string {
  const userMessages = messages.filter((msg) => msg.role === "user").map((msg) => normalizeText(msg.content));
  const assistantMessages = messages
    .filter((msg) => msg.role === "assistant")
    .map((msg) => normalizeText(msg.content));

  const userHead = userMessages[0] ?? "";
  const userTail = userMessages[userMessages.length - 1] ?? "";
  const assistantTail = assistantMessages[assistantMessages.length - 1] ?? "";

  const parts = [
    userHead ? `用户提到：${truncate(userHead, 80)}` : "",
    userTail && userTail !== userHead ? `后续重点：${truncate(userTail, 80)}` : "",
    assistantTail ? `助手响应：${truncate(assistantTail, 80)}` : "",
  ].filter(Boolean);

  if (parts.length === 0) return "该窗口没有可用文本，跳过结构化摘要。";
  return parts.join("；");
}

export function buildSituationTimeInfo(timestamp: string, summary: string): string {
  const date = new Date(timestamp);
  if (Number.isNaN(date.getTime())) return `未知时间场景：${truncate(summary, 120)}`;
  const hour = String(date.getHours()).padStart(2, "0");
  const minute = String(date.getMinutes()).padStart(2, "0");
  return `${date.toISOString().slice(0, 10)} ${hour}:${minute} 用户正在推进：${truncate(summary, 120)}`;
}
